### ISSUE TITLE
Image Upload Error

### ISSUE THREAD

**Issue by DhruveshTripathi** [Coyote6/drupal-7-grapesjs#1]
I installed the module with all the necessary libraries. It seems to be working fine except one issue. When the form get initialized, I get the following error:

There was a problem initializing the upload form. You may be unable to upload images using the GrapesJs editor.

Can you please provide the solution?
Thanks

**Comment by Coyote6** [comment]
Hi I just uploaded my most recent version of the module.  You may give it a whirl.  If you still get the error, please let me know.  It is caused, when the javascript is unable to retrieve the form values for the upload form.  Let me know if you still get the error.

**Comment by DhruveshTripathi** [comment]
Hello @Coyote6 
I just tried the latest version of the module and got the following error for Image Upload:

"There was a problem initializing the upload form. You may be unable to upload images using the GrapesJs editor."

I see some improvements like when I give web url to upload iamge, it gets loaded. But Local images can't be uploaded.

Also, while adding text. cursor position resets to the beginning of the line.

Improvements are very good from the last release. Good going!
Thanks!

**Comment by Coyote6** [comment]
Hi @DhruveshTripathi,

I think I may have found your issue.  Make sure you give your user not only access to use the "Visual Editor" text format, but there is also a permission under the "GrapeJs" section for using the Editor.  Make sure your user has both.  This definitely needs to be cleaned up on my end because the JavaScripts were never supposed to attach without the "GrapesJs" permission. 

If you can, see if that works for you.  Thanks.

**Comment by DhruveshTripathi** [comment]
Hey @Coyote6 

I assigned the grapesJS permission to my user role but still I no success. Even without assigning permission, user role can access grapesJS editor if it has permission to access "Visual Editor".

**Comment by Coyote6** [comment]
Yes I noticed they had access yesterday while testing it. The JavaScripts were not meant to be added to the page but are. It is a bug that needs to be fixed. I still do not know why you are getting the error though... I tested on a clean install and once the permission was added the alert went away.  Only other thing I can think of right off hand is that something is blocking the ajax request to get the form.  Can you please see if any errors are appearing in your web console? And screenshot them if there are any.  Thanks.

**Comment by DhruveshTripathi** [comment]
Hey @Coyote6 

Sorry I forgot to mention that the alert "There was a problem initializing the upload form. You may be unable to upload images using the GrapesJs editor." went away after granting permission to user role but still user is not able to upload Images from local.

**Comment by Coyote6** [comment]
Are you dragging an image onto the canvas first and then dragging the image into the uploader or just trying to drag the image straight to the canvas?

**Comment by DhruveshTripathi** [comment]
@Coyote6 

After placing Image block on canvas, when I try to upload Image from local by Drag and Drop or by clicking on "Drop Files Here or click to upload", it doesn't work. But when I provide local Image's path in Address bar next to "Add Image" button, it works.

**Comment by Coyote6** [comment]
Is your site being hosted locally as well or is it on a remote server?

**Comment by DhruveshTripathi** [comment]
I'm running this on my local right now. But my website is live, and I'm trying to integrate grapesJS with Drupal. If I get success, I'm planning it to push it on my remote server.

**Comment by Coyote6** [comment]
Ok let me test on my local machine today. All my testing has been remote... we are working on making it a contributed module right now... future home will be at https://www.drupal.org/project/grapesjs

**Comment by Coyote6** [comment]
Are you running clean urls currently? All paths are hard coded at the moment to use clean urls.

**Comment by DhruveshTripathi** [comment]
Yes. I'm running on clean URLs. Actually I just want to provide user grapesJS in my running project, so I'm trying to integrate. Glad to know that, development is going well for making it full contributed module on Drupal.org. Will definitely try to contribute :)

**Comment by Coyote6** [comment]
Okay... well I think I have tracked down the issue at least for my environment and was successfully able to upload images locally.  Here are what my issues were:

Issue #1:
My clean urls were not actually working properly.  I could access the page via the direct root, but clean urls would not validate.  I manually changed the setting to turn them on, but that did not work as any form I submitted erred due to drupal setting a 404 on the page even though it existed and could be viewed.  (More about this issue here: https://www.drupal.org/project/documentation/issues/1178850)

Basically even though mod rewrite was working something in Drupal did not recognize it as working. Thus breaking any form I tried to submit, grapejs or not.

So I shut clean urls back off and the forms started functioning once more.  So to fix the issue, modify the editor.js file to use non-clean urls... 
Changing the urls like /editor/grapesjs/upload into /?q=editor/grapesjs/upload

I am currently working to make this into a proper setting that runs off the clean url variable.

Issue #2:
Once I got that working, I realized that the files erred on upload due to the nid not being set in the hook_entity_presave() function.  Because it is a new node it obviously shouldn't be set yet due to it being new and it hasn't been saved.  So I added a hook_entity_insert() to recall the hook_entity_presave() function once it has the id, and if no id is set, then it skips to prevent errors.

Once I get issue #1 running properly I will recommit the updates.

**Comment by DhruveshTripathi** [comment]
Awesome! Thanks a lot! I'll try it on my local again.
